### PR TITLE
Add sparse concat operator support

### PR DESCRIPTION
**Pull request by kalyc**
### Summary
Add sparse support for `concat` operator

### Related Issues
[Missing sparse operators](https://github.com/awslabs/keras-apache-mxnet/issues/18)

### PR Overview

- [x] This PR requires new unit tests [y/n] (make sure tests are included)
- [ ] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [x] This PR is backwards compatible [y/n]
- [ ] This PR changes the current API [y/n]

### CODE REVIEW

File: keras/backend/mxnet_backend.py [hunk 1]
```diff
@@ -2038,8 +2044,12 @@ def concatenate(tensors, axis=-1):
         else:
             axis = 0
 
-    tensors = [t.symbol for t in tensors]
-    return KerasSymbol(mx.sym.concat(*tensors, dim=axis))
+    symbols = [t.symbol for t in tensors]
+
+    if py_all([is_sparse(t) for t in tensors]):
+        return KerasSymbol(mx.sym.sparse.concat(*symbols, dim=0))
```
Comment (roywei) [inline keras/backend/mxnet_backend.py]: you should still check if axis !=0, even if all tensors are sparse, if user want to concat on axis other than 0, we use dense concat

File: keras/backend/mxnet_backend.py [hunk 2]
```diff
@@ -2038,8 +2044,12 @@ def concatenate(tensors, axis=-1):
...
+
+    return KerasSymbol(mx.sym.concat(*symbols, dim=axis))
```
Comment (roywei) [inline keras/backend/mxnet_backend.py]: If part of tensors are sparse, so it will fail the check at `py_all([is_sparse(t) for t in tensors]):`, what will happen if you pass them to dense concat?

File: tests/keras/backend/mxnet_sparse_test.py [hunk 3]
```diff
@@ -104,6 +104,35 @@ def test_sparse_dot(self):
         assert k_s.shape == k_d.shape
         assert_allclose(k_s, k_d, atol=1e-05)
 
+    def test_sparse_concat(self):
+        x_d = np.array([0, 7, 2, 3], dtype=np.float32)
```
Comment (roywei) [inline tests/keras/backend/mxnet_sparse_test.py]: please use sparse data generator

File: keras/backend/mxnet_backend.py [hunk 4]
```diff
@@ -17,6 +17,8 @@
 _REENTRY = False
 NAME_SCOPE_STACK = []
 
+py_all = all
```
Comment (roywei) [inline keras/backend/mxnet_backend.py]: why do we need to define py_all here?

Comment (kalyc) [reply keras/backend/mxnet_backend.py]: following the convention in `tensorflow_backend` file

File: tests/keras/backend/mxnet_sparse_test.py (hunk 3 above)
Comment (kalyc) [reply tests/keras/backend/mxnet_sparse_test.py]: done

File: keras/backend/mxnet_backend.py (hunk 1 above)
Comment (kalyc) [reply keras/backend/mxnet_backend.py]: we have hard-coded the dimension - why do we need to check the axis!=0 here?

Comment (kalyc) [reply keras/backend/mxnet_backend.py]: removed hard-coded value & added unit test

File: keras/backend/mxnet_backend.py [hunk 5]
```diff
@@ -2038,8 +2044,12 @@ def concatenate(tensors, axis=-1):
         else:
             axis = 0
 
-    tensors = [t.symbol for t in tensors]
-    return KerasSymbol(mx.sym.concat(*tensors, dim=axis))
+    symbols = [t.symbol for t in tensors]
+
+    if axis == 0 and py_all([is_sparse(t) for t in tensors]):
```
Comment (sandeep-krishnamurthy) [inline keras/backend/mxnet_backend.py]: nit: it is preferrable to avoid multiple returns in a func.

File: keras/backend/mxnet_backend.py [hunk 6]
```diff
@@ -2038,8 +2044,12 @@ def concatenate(tensors, axis=-1):
...
+        return KerasSymbol(mx.sym.sparse.concat(*symbols, dim=axis))
+
+    return KerasSymbol(mx.sym.concat(*symbols, dim=axis))
```
Comment (sandeep-krishnamurthy) [inline keras/backend/mxnet_backend.py]: We should show warning if axis !=0 and is_sparse(t) for t in tensors? I am assuming you do dense concat automatically if axis !=0. This has performance implication and important for users to know.

Comment (kalyc) [reply keras/backend/mxnet_backend.py]: this is replicating the behavior for sparse concat in mxnet - http://mxnet.apache.org/api/python/symbol/symbol.html#mxnet.symbol.concat - do we still want to add a warning there?

Comment (sandeep-krishnamurthy) [reply keras/backend/mxnet_backend.py]: I think yes. Users should know that tensor is being converted to dense.
